File: domain/src/test/java/com/kaciula/archiman/domain/usecases/InitColdStartTest.java
package com.kaciula.archiman.domain.usecases;


import com.kaciula.archiman.domain.boundary.data.AppInfoRepository;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;

public class InitColdStartTest {

  @Mock AppInfoRepository appInfoRepository;

  private InitColdStart useCase;

  @Before
  public void setupUseCase() {
    MockitoAnnotations.initMocks(this);

    useCase = new InitColdStart(appInfoRepository);
  }

  @Test
  public void firstTime() {
    int currentVersionCode = 1;
    Mockito.when(appInfoRepository.isFirstTime()).thenReturn(true);
    Mockito.when(appInfoRepository.getVersionCode()).thenReturn(0);

    useCase.execute(InitColdStart.RequestModel.create(currentVersionCode))
        .subscribe(new Observer<InitColdStart.ResponseModel>() {
                     @Override
                     public void onSubscribe(Disposable d) {
                       // do nothing
                     }

                     @Override
                     public void onNext(InitColdStart.ResponseModel value) {
                       Assert.fail("onNext should not be triggered");
                     }

                     @Override
                     public void onError(Throwable e) {
                       Assert.fail("Failure to init from cold start");
                     }

                     @Override
                     public void onComplete() {
                       // do nothing
                     }
                   }
        );

    Mockito.verify(appInfoRepository).saveFirstTime(false);
    Mockito.verify(appInfoRepository).saveVersionCode(currentVersionCode);
  }

  @Test
  public void newVersion() {
    int currentVersionCode = 2;
    Mockito.when(appInfoRepository.isFirstTime()).thenReturn(false);
    Mockito.when(appInfoRepository.getVersionCode()).thenReturn(1);

    useCase.execute(InitColdStart.RequestModel.create(currentVersionCode))
        .subscribe(new Observer<InitColdStart.ResponseModel>() {
                     @Override
                     public void onSubscribe(Disposable d) {
                       // do nothing
                     }

                     @Override
                     public void onNext(InitColdStart.ResponseModel value) {
                       Assert.fail("onNext should not be triggered");
                     }

                     @Override
                     public void onError(Throwable e) {
                       Assert.fail("Failure to init from cold start");
                     }

                     @Override
                     public void onComplete() {
                       // do nothing
                     }
                   }
        );

    Mockito.verify(appInfoRepository).saveVersionCode(currentVersionCode);
  }
}

File: domain/src/test/kotlin/com/kaciula/archiman/domain/usecases/InitColdStartTest.kt
package com.kaciula.archiman.domain.usecases


import com.kaciula.archiman.domain.boundary.infrastructure.AppRepository
import com.kaciula.archiman.domain.model.App
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations

class InitColdStartTest {

    @Mock
    private lateinit var appRepository: AppRepository

    private lateinit var useCase: InitColdStart

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)

        useCase = InitColdStart(appRepository)
    }

    @Test
    fun firstTime() {
        val appBefore = App(true, 0)
        val currentVersionCode = 1
        val appAfter = App(false, currentVersionCode)
        Mockito.`when`(appRepository.get()).thenReturn(appBefore)

        val observer = useCase.execute(InitColdStart.RequestModel(currentVersionCode)).test()

        observer.assertNoErrors()
        Mockito.verify<AppRepository>(appRepository).save(appAfter)
    }

    @Test
    fun newVersion() {
        val appBefore = App(false, 1)
        val currentVersionCode = 2
        val appAfter = App(false, currentVersionCode)
        Mockito.`when`(appRepository.get()).thenReturn(appBefore)

        val observer = useCase.execute(InitColdStart.RequestModel(currentVersionCode)).test()

        observer.assertNoErrors()
        Mockito.verify<AppRepository>(appRepository).save(appAfter)
    }
}
